applications/FluidDynamicsApplication/custom_elements/two_fluid_navier_stokes.cpp

```diff
@@ -0,0 +1,2054 @@
+            GeometryType::ShapeFunctionsGradientsType shape_derivatives_enr_pos;
+            GeometryType::ShapeFunctionsGradientsType shape_derivatives_enr_neg;
+            ComputeSplitting(
+                data,
+                shape_functions_pos,
+                shape_functions_neg,
+                shape_functions_enr_pos,
+                shape_functions_enr_neg,
+                shape_derivatives_pos,
+			    shape_derivatives_neg,
+                shape_derivatives_enr_pos,
+                shape_derivatives_enr_neg);
+
+			if (data.NumberOfDivisions == 1) {
+				//cases exist when the element is not subdivided due to the characteristics of the provided distance
+				//in this cases the element is treated as AIR or FLUID depending on the side
+				Vector gauss_weights; 
+                Matrix shape_functions; 
+                ShapeFunctionDerivativesArrayType shape_derivatives; 
+                this->CalculateGeometryData( 
+                    gauss_weights, shape_functions, shape_derivatives); 
+                const unsigned int number_of_gauss_points = gauss_weights.size(); 
+				array_1d<double,NumNodes> Ncenter;
+				for(unsigned int i=0; i<NumNodes; i++) Ncenter[i]=0.25;
+				const double dgauss = inner_prod(data.Distance, Ncenter);
+				for (unsigned int g = 0; g < number_of_gauss_points; g++) {
+					data.UpdateGeometryValues(g,gauss_weights[g], row(shape_functions, g),
+						shape_derivatives[g]);
+
+                    data.CalculateDensityAtGaussPoint();
+
+					if (dgauss > 0.0)
+						data.CalculateAirMaterialResponse();
+					else
+						this->CalculateMaterialResponse(data);				
+
+					this->AddTimeIntegratedSystem(
+						data, rLeftHandSideMatrix, rRightHandSideVector);
+				}
+			}
+
+			else {
+				MatrixType Vtot;
+				MatrixType Htot;
+				MatrixType Kee_tot;
+				VectorType rhs_ee_tot;
+				Vtot.resize(NumNodes*(Dim + 1), NumNodes, false);
+				Htot.resize(NumNodes, NumNodes*(Dim + 1), false);
+				Kee_tot.resize(NumNodes, NumNodes, false);
+				rhs_ee_tot.resize(NumNodes, false);
+
+                noalias(Vtot) = ZeroMatrix(NumNodes*(Dim + 1), NumNodes);
+                noalias(Htot) = ZeroMatrix(NumNodes, NumNodes*(Dim + 1));
+                noalias(Kee_tot) = ZeroMatrix(NumNodes, NumNodes);
+                noalias(rhs_ee_tot) = ZeroVector(NumNodes);
+
+                for (unsigned int g_pos = 0; g_pos < data.w_gauss_pos_side.size(); g_pos ++) {
+                    data.UpdateGeometryValues(
+                        g_pos,
+                        data.w_gauss_pos_side[g_pos],
+                        row(shape_functions_pos, g_pos),
+				        shape_derivatives_pos[g_pos],
+                        row(shape_functions_enr_pos,g_pos),
+                        shape_derivatives_enr_pos[g_pos]);
+                    
+                    data.CalculateDensityAtGaussPoint();
+                    data.CalculateAirMaterialResponse();
```

Comment:
I don't really like that you are adding new methods to your data, since this defeats what we wanted to achieve with templating (you would like not to care about the detais of your data, but now you need to be sure that data implements these specific functions, which are particular to your implementation).

I am also aware that I have the same problem in the base element (and it would be smarter to move everything into UpdateGeometryValues, maybe changing its name), so I will not block the PR for this, but just be aware that we might rewrite this part in the future.